src/main/java/io/numaproj/numaflow/mapper/Service.java

```diff
@@ -1,59 +1,90 @@
 package io.numaproj.numaflow.mapper;
 
-import com.google.protobuf.ByteString;
+import akka.actor.ActorRef;
+import akka.actor.ActorSystem;
 import com.google.protobuf.Empty;
+import io.grpc.Status;
 import io.grpc.stub.StreamObserver;
 import io.numaproj.numaflow.map.v1.MapGrpc;
 import io.numaproj.numaflow.map.v1.MapOuterClass;
 import lombok.AllArgsConstructor;
 import lombok.extern.slf4j.Slf4j;
 
-import java.time.Instant;
-import java.util.ArrayList;
-import java.util.List;
+import java.util.concurrent.CompletableFuture;
 
 import static io.numaproj.numaflow.map.v1.MapGrpc.getMapFnMethod;
 
 @Slf4j
 @AllArgsConstructor
 class Service extends MapGrpc.MapImplBase {
 
+    public static final ActorSystem mapperActorSystem = ActorSystem.create("mapper");
+
     private final Mapper mapper;
 
-    /**
-     * Applies a function to each datum element.
-     */
+    // TODO we need to propagate the exception all the way up and shutdown the server.
+    static void handleFailure(
+            CompletableFuture<Void> failureFuture) {
+        new Thread(() -> {
+            try {
+                failureFuture.get();
+            } catch (Exception e) {
+                e.printStackTrace();
+            }
+        }).start();
+    }
+
     @Override
-    public void mapFn(
-            MapOuterClass.MapRequest request,
-            StreamObserver<MapOuterClass.MapResponse> responseObserver) {
+    public StreamObserver<MapOuterClass.MapRequest> mapFn(final StreamObserver<MapOuterClass.MapResponse> responseObserver) {
 
         if (this.mapper == null) {
-            io.grpc.stub.ServerCalls.asyncUnimplementedUnaryCall(
+            return io.grpc.stub.ServerCalls.asyncUnimplementedStreamingCall(
                     getMapFnMethod(),
                     responseObserver);
-            return;
         }
 
-        Datum handlerDatum = new HandlerDatum(
-                request.getValue().toByteArray(),
-                Instant.ofEpochSecond(
-                        request.getWatermark().getSeconds(),
-                        request.getWatermark().getNanos()),
-                Instant.ofEpochSecond(
-                        request.getEventTime().getSeconds(),
-                        request.getEventTime().getNanos()),
-                request.getHeadersMap()
-        );
-
-        // process request
-        MessageList messageList = mapper.processMessage(request
-                .getKeysList()
-                .toArray(new String[0]), handlerDatum);
-
-        // set response
-        responseObserver.onNext(buildResponse(messageList));
-        responseObserver.onCompleted();
+        CompletableFuture<Void> failureFuture = new CompletableFuture<>();
+
+        handleFailure(failureFuture);
```

Comment:
why do we handle a failure future right after creating it?

Reply:
we handle failure using a separate thread.